src/cpu/simple/probes/looppoint_analysis.hh

```diff
@@ -0,0 +1,331 @@
+     * compare with the regionLength to determine if we reach the end of a
+     * region.
+     */
+    uint64_t globalInstCounter;
+
+    /**
+     * This variable stores the Program Counter address of the most recent
+     * valid backward branch that we consider as a marker.
+     */
+    Addr mostRecentBackwardBranchPC;
+
+    /**
+     * This set stores the Program Counter addresses of the valid backward
+     * branches.
+     */
+    std::unordered_set<Addr> backwardBranchPC;
+    /**
+     * This set stores the Program Counter addresses of the valid not control
+     * instructions.
+     */
+    std::unordered_set<Addr> validNotControlPC;
+    /**
+     * This set stores the Program Counter addresses of the valid control
+     * instructions.
+     */
+    std::unordered_set<Addr> validControlPC;
+    /**
+     * This set stores the Program Counter addresses of the encountered
+     * instructions.
+     */
+    std::unordered_set<Addr> encounteredPC;
+
+  public:
+    bool ifBackwardBranch(const Addr pc) const
+    {
+        return backwardBranchPC.find(pc) != backwardBranchPC.end();
+    };
+
+    bool ifValidNotControl(const Addr pc) const
+    {
+        return validNotControlPC.find(pc) != validNotControlPC.end();
+    };
+
+    bool ifValidControl(const Addr pc) const
+    {
+        return validControlPC.find(pc) != validControlPC.end();
+    };
+
+    bool ifEncountered(const Addr pc) const
+    {
+        return encounteredPC.find(pc) != encounteredPC.end();
+    };
+
+    void updateBackwardBranch(const Addr pc)
+    {
+        backwardBranchPC.insert(pc);
+    };
+
+    void updateValidNotControl(const Addr pc)
+    {
+        validNotControlPC.insert(pc);
+    };
+
+    void updateValidControl(const Addr pc)
+    {
+        validControlPC.insert(pc);
+    };
+
+    void updateEncountered(const Addr pc)
+    {
+        encounteredPC.insert(pc);
+    };
+
+    void updateBBInstMap(Addr pc, uint64_t inst_ount)
+    {
+        if (bbInstMap.find(pc) == bbInstMap.end())
+        {
+            bbInstMap.insert(std::make_pair(pc, inst_ount));
+        }
+    };
+
+    std::unordered_map<Addr, uint64_t> getGlobalBBV() const
+    {
+        return globalBBV;
+    };
+
+    void clearGlobalBBV()
+    {
+        globalBBV.clear();
+        DPRINTF(LooppointAnalysis,"globalBBV is cleared\n");
+    };
+
+    uint64_t getGlobalInstCounter() const
+    {
+        return globalInstCounter;
+    };
+
+    void clearGlobalInstCounter()
+    {
+        globalInstCounter = 0;
+        DPRINTF(LooppointAnalysis,"globalInstCounter is cleared\n current "
+            "globalInstCounter = %lu\n", globalInstCounter);
+    };
+
+    void incrementGlobalInstCounter()
+    {
+        globalInstCounter++;
+    };
+
+    Addr getMostRecentBackwardBranchPC() const
+    {
+        return mostRecentBackwardBranchPC;
+    };
+
+    std::unordered_map<Addr, uint64_t> getBackwardBranchCounter() const
```

Comment:
We should be careful with this function. This map could be very large and this function will likely return a copy. Where is it called? Should it return a copy? should it return a reference?

Reply:
It can be called at every end of the region. In the Python script, the map will be stored as a dictionary. We don't expect to change anything in the map. 
I think by default, Pybind converts it to dictionary that copied the map's data, so it doesn't return a reference, but since we don't expect to change anything in the map. returning a copy should be fine.